x-pack/filebeat/input/gcs/job.go

```diff
@@ -121,57 +131,207 @@ func (j *job) Timestamp() time.Time {
+		}
+		if j.isRootArray {
+			j.state.setRootArray(j.object.Name)
+		}
+	}
+	dec := json.NewDecoder(r)
+	// If array is present at root then read json token and advance decoder
+	if j.isRootArray {
+		_, err := dec.Token()
+		if err != nil {
+			return fmt.Errorf("failed to read JSON token for object: %s, with error: %w", j.object.Name, err)
+		}
+	}
+
+	var offset, relativeOffset int64
+	// uncompressed files use the client to directly set the offset, this
+	// in turn causes the offset to reset to 0 for the new stream, hence why
+	// we need to keep relative offsets to keep track of the actual offset
+	if !j.isCompressed && !j.isRootArray {
+		relativeOffset = j.offset
+	}
+	dec.UseNumber()
+	for dec.More() && ctx.Err() == nil {
+		var item json.RawMessage
+		offset = dec.InputOffset()
+		if err = dec.Decode(&item); err != nil {
+			return fmt.Errorf("failed to decode json: %w", err)
+		}
+		// manually seek offset only if file is compressed or if root element is an array
+		if (j.isCompressed || j.isRootArray) && offset < j.offset {
+			continue
+		}
+		data := []byte(item)
+		var parsedData []mapstr.M
+		if j.src.ParseJSON {
+			reader := bytes.NewReader(data)
+			parsedData, err = decodeJSON(reader)
+			if err != nil {
+				j.log.Errorf(jobErrString, id, err)
+			}
+		}
+		evt := j.createEvent(data, parsedData, offset+relativeOffset)
+		// updates the offset after reading the file
+		// this avoids duplicates for the last read when resuming operation
+		offset = dec.InputOffset()
+		if !dec.More() {
+			// if this is the last object, then peform a complete state save
+			j.state.save(j.object.Name, j.object.Updated)
+		} else {
+			// partially saves read state using offset
+			j.state.savePartial(j.object.Name, offset+relativeOffset)
+		}
+		if err := j.publisher.Publish(evt, j.state.checkpoint()); err != nil {
+			j.log.Errorf(jobErrString, id, err)
+		}
+	}
+	return nil
+}
+
+// addGzipDecoderIfNeeded determines whether the given stream of bytes (encapsulated in a buffered reader)
+// represents gzipped content or not and adds gzipped decoder if needed. A bufio.Reader is used
+// so the function can peek into the byte  stream without consuming it. This makes it convenient for
+// code executed after this function call to consume the stream if it wants.
+func (j *job) addGzipDecoderIfNeeded(reader *bufio.Reader) (io.Reader, bool, error) {
+	isStreamGzipped := false
+	// check if stream is gziped or not
+	buf, err := reader.Peek(3)
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			err = nil
+		}
+		return reader, false, err
+	}
+
+	// gzip magic number (1f 8b) and the compression method (08 for DEFLATE).
+	isStreamGzipped = bytes.Equal(buf, []byte{0x1F, 0x8B, 0x08})
 
-	_, err = data.ReadFrom(reader)
+	if !isStreamGzipped {
+		return reader, false, nil
+	}
+
+	gzReader, err := gzip.NewReader(reader)
 	if err != nil {
-		return nil, fmt.Errorf("failed to read data from object reader with error: %w", err)
+		return reader, false, err
```

Comment:
```suggestion
		return nil, false, err
```
If `gzip.NewReader` has failed, it will have consumed some bytes of `reader`, so it is no longer in a state that we know we can use and should be dropped defensively.